File: packages/server-client/src/parquet-stream-shapes.ts
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

/**
 * The client half of cross-batch shape sharing on
 * `POST /api/v1/parse/parquet-stream?stream_shapes=cross-batch` (issue #5407).
 *
 * On that stream a batch carries only the shapes no earlier batch did, and its
 * mesh rows index the vertex and index rows of the WHOLE stream so far. So the
 * reader keeps every batch's vertex and index columns, appended in order, and
 * builds each batch's meshes against all of them. What it keeps is the
 * stream's distinct shape geometry, sent once, which is what makes the stream
 * small in the first place; occurrences are materialized per batch and handed
 * off exactly as before.
 *
 * Each batch states where its tables start (`vertex_base`, `index_base`). The
 * store refuses a batch whose base is not what it holds, so a dropped or
 * reordered batch fails loud instead of drawing one shape's vertices for
 * another. A batch WITHOUT bases came from a server that did not honour the
 * opt-in and decodes on its own, as every batch did before.
 */

import type { MeshData } from './types.js';
import { numericColumn } from './parquet-columns.js';
import { readFlatTables } from './parquet-decoder.js';
import { buildMeshesFromTables, type ArrowTableLike } from './parquet-tables.js';

const VERTEX_COLUMNS = ['x', 'y', 'z', 'nx', 'ny', 'nz'] as const;
const INDEX_COLUMNS = ['i0', 'i1', 'i2'] as const;

/** An append-only typed column that doubles its capacity as it grows. */
class GrowableColumn {
  private buffer: Float32Array | Uint32Array;
  private length = 0;

  constructor(private readonly allocate: (capacity: number) => Float32Array | Uint32Array) {
    this.buffer = allocate(1024);
  }

  append(values: ArrayLike<number>): void {
    const needed = this.length + values.length;
    if (needed > this.buffer.length) {
      const grown = this.allocate(Math.max(needed, this.buffer.length * 2));
      grown.set(this.buffer.subarray(0, this.length));
      this.buffer = grown;
    }
    this.buffer.set(values, this.length);
    this.length = needed;
  }

  view(): Float32Array | Uint32Array {
    return this.buffer.subarray(0, this.length);
  }
}

function columns(names: readonly string[], allocate: (n: number) => Float32Array | Uint32Array) {
  return new Map(names.map((name) => [name, new GrowableColumn(allocate)]));
}

function tableOf(cols: Map<string, GrowableColumn>): ArrowTableLike {
  return {
    getChild(name) {
      const view = cols.get(name)?.view();
      return view ? { toArray: () => view, get: (i) => view[i] } : null;
    },
  };
}

/** Every vertex and index row a cross-batch stream has sent so far. */
export class StreamShapeStore {
  private readonly vertex = columns(VERTEX_COLUMNS, (n) => new Float32Array(n));
  private readonly index = columns(INDEX_COLUMNS, (n) => new Uint32Array(n));
  /** Vertex rows received. */
  private vertices = 0;
  /** Indices received, three per index-table row: the unit of `index_base`. */
  private indices = 0;

  /**
   * Append one batch's vertex and index tables at the bases the batch states.
   * Throws, appending nothing, when a base is not what the store holds or a
   * table is malformed.
   */
  append(vertexArrow: ArrowTableLike, indexArrow: ArrowTableLike, vertexBase: number, indexBase: number): void {
    if (vertexBase !== this.vertices || indexBase !== this.indices) {
      throw new Error(
        `Malformed Parquet stream: batch starts at vertex ${vertexBase} / index ${indexBase}, ` +
          `but ${this.vertices} / ${this.indices} were received (a batch is missing or out of order)`
      );
    }
    const vertexCols = parallelColumns(vertexArrow, VERTEX_COLUMNS);
    const indexCols = parallelColumns(indexArrow, INDEX_COLUMNS);
    VERTEX_COLUMNS.forEach((name, i) => this.vertex.get(name)?.append(vertexCols[i]));
    INDEX_COLUMNS.forEach((name, i) => this.index.get(name)?.append(indexCols[i]));
    this.vertices += vertexCols[0].length;
    this.indices += 3 * indexCols[0].length;
  }

  vertexTable(): ArrowTableLike {
    return tableOf(this.vertex);
  }

  indexTable(): ArrowTableLike {
    return tableOf(this.index);
  }
}

/** All of `names`, present and of one length, or throw. */
function parallelColumns(table: ArrowTableLike, names: readonly string[]): ArrayLike<number>[] {
  const cols: ArrayLike<number>[] = [];
  for (const name of names) {
    const col = numericColumn(table, name);
    if (!col || (cols.length > 0 && col.length !== cols[0].length)) {
      throw new Error(`Malformed Parquet stream: missing or ragged ${names.join('/')} columns`);
    }
    cols.push(col);
  }
  return cols;
}

/**
 * Decode one cross-batch stream batch against everything the stream has sent:
 * append its tables to `store`, then build its meshes, whose ranges may reach
 * back into any earlier batch.
 */
export async function decodeCrossBatch(
  data: ArrayBuffer,
  store: StreamShapeStore,
  vertexBase: number,
  indexBase: number
): Promise<MeshData[]> {
  const { meshArrow, vertexArrow, indexArrow } = await readFlatTables(data);
  store.append(vertexArrow, indexArrow, vertexBase, indexBase);
  return buildMeshesFromTables(meshArrow, store.vertexTable(), store.indexTable());
}

File: apps/server/src/services/parquet_stream_shapes.rs
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

//! Whole-stream offsets for the streaming Parquet route, and the cross-batch
//! shape sharing that rides on them (#5407).
//!
//! A streamed model is written one batch at a time, so anything that has to
//! hold across batches lives here: the running vertex/index offsets every
//! batch's mesh table is based on, and, when the client opted in, where each
//! distinct shape already landed. The cache writer and the per-batch client
//! blob both consume a [`PlannedBatch`], so the two can never disagree about a
//! range.
//!
//! BOUNDED MEMORY. No occurrence outlives the batch that produced it. What
//! grows across batches is, per DISTINCT shape emitted so far, one
//! [`EmittedShape`] (16 bytes) plus its `MeshGeometryKey` (24 bytes), whatever
//! the size of that shape's geometry; and, per instanced representation, the
//! ONE template mesh its later occurrences are verified against, capped in
//! total at [`RETAINED_TEMPLATE_BUDGET_BYTES`]. Past the cap a representation
//! is simply not carried: its later occurrences are shared within their own
//! batch, or by content hash, instead. Sharing degrades; correctness and the
//! bound do not.

use crate::services::parquet::{frame_sections, write_parquet_buffer, ParquetError};
use crate::services::parquet_layout::{ParquetLayout, StreamShapes};
use crate::services::parquet_mesh_tables::build_mesh_tables;
use crate::services::parquet_shape_plan::{EmittedShape, PriorShapes, RetainedTemplate, ShapePlan};
use crate::types::MeshData;
use arrow::record_batch::RecordBatch;
use bytes::Bytes;

/// Ceiling on the template geometry a cross-batch stream retains (#5407).
///
/// Sized against what it buys, not what it costs a small model: the
/// representations that repeat across batches are the model's distinct
/// instanced shapes, whose geometry is what the shared stream sends ONCE
/// anyway. 256 MiB of raw template buffers is several times that on every
/// fixture measured for #5407, while staying well under one parse's working
/// set on the models a streaming deployment exists for.
pub(super) const RETAINED_TEMPLATE_BUDGET_BYTES: usize = 256 * 1024 * 1024;

/// One batch's mesh / vertex / index tables, planned against the whole stream.
pub struct PlannedBatch {
    pub(super) mesh: RecordBatch,
    pub(super) vertex: RecordBatch,
    pub(super) index: RecordBatch,
    /// Whole-stream offset of this batch's first vertex row.
    pub vertex_base: u32,
    /// Whole-stream offset of this batch's first index, in indices (three per
    /// index-table row), the unit of the mesh table's `index_start`.
    pub index_base: u32,
}

impl PlannedBatch {
    /// The batch as a client blob, in the same `[len][mesh][len][vert][len]
    /// [idx]` framing every batch event carries. Its offsets are whole-stream,
    /// so it is only meaningful to a client told `vertex_base` / `index_base`.
    pub fn encode(&self) -> Result<Bytes, ParquetError> {
        frame_sections(
            &write_parquet_buffer(&self.mesh)?,
            &write_parquet_buffer(&self.vertex)?,
            &write_parquet_buffer(&self.index)?,
        )
    }

    /// Mesh rows in this batch.
    pub fn mesh_count(&self) -> usize {
        self.mesh.num_rows()
    }
}

/// The running state behind a streamed model's whole-stream offsets.
pub struct StreamShapePlanner {
    layout: ParquetLayout,
    /// `None`: every mesh emits its own block, as the stream always did.
    /// `Some`: cross-batch sharing, against every shape emitted so far.
    prior: Option<PriorShapes>,
    /// Raw buffer bytes held by `prior`'s retained templates, and their cap.
    retained_bytes: usize,
    template_budget: usize,
    vertex_offset: u32,
    index_offset: u32,
}

impl StreamShapePlanner {
    /// A planner for `layout`. Sharing needs BOTH the shared-shapes layout and
    /// the client's cross-batch opt-in; the route rejects the opt-in on the
    /// flat layout, so ignoring it here is never what a caller sees.
    pub fn new(layout: ParquetLayout, stream_shapes: StreamShapes) -> Self {
        let sharing = layout.has_rotation() && stream_shapes == StreamShapes::CrossBatch;
        Self {
            layout,
            prior: sharing.then(PriorShapes::default),
            retained_bytes: 0,
            template_budget: RETAINED_TEMPLATE_BUDGET_BYTES,
            vertex_offset: 0,
            index_offset: 0,
        }
    }

    /// Plan the next batch and advance the whole-stream offsets past the
    /// shapes it emits. Fails loud, advancing nothing, when an offset would
    /// overflow the mesh table's `u32` columns: wrapping would decode as
    /// garbage, not as an error.
    ///
    /// `baked_basis` is the row-major `native_to_baked` of the frame the
    /// batch's vertices are in (`StreamEvent::Batch::baked_basis`); ignored
    /// unless the planner shares.
    pub fn plan(
        &mut self,
        meshes: &[MeshData],
        baked_basis: Option<&[f64; 16]>,
    ) -> Result<PlannedBatch, ParquetError> {
        let basis = baked_basis.map(|m| ifc_lite_geometry::Matrix4::from_row_slice(m));
        let (plan, new_shapes) = match &self.prior {
            Some(prior) => ShapePlan::shared_shapes_after(meshes, basis.as_ref(), prior),
            None => (ShapePlan::Identity, Vec::new()),
        };
        let (vertex_base, index_base) = (self.vertex_offset, self.index_offset);
        let mut emitted = Vec::with_capacity(new_shapes.len());
        let (mut vertex_end, mut index_end) = (vertex_base, index_base);
        for shape in plan.shape_meshes(meshes) {
            let vertex_count = u32::try_from(shape.positions.len() / 3).ok();
            let index_count = u32::try_from(shape.indices.len()).ok();
            let (Some(vertex_count), Some(index_count)) = (vertex_count, index_count) else {
                return Err(overflow());
            };
            if self.prior.is_some() {
                emitted.push(EmittedShape {
                    vertex_start: vertex_end,
                    vertex_count,
                    index_start: index_end,
                    index_count,
                });
            }
            vertex_end = vertex_end.checked_add(vertex_count).ok_or_else(overflow)?;
            index_end = index_end.checked_add(index_count).ok_or_else(overflow)?;
        }
        let (mesh, vertex, index) =
            build_mesh_tables(meshes, &plan, self.layout, vertex_base, index_base)?;
        if let Some(prior) = self.prior.as_mut() {
            // `new_shapes` runs parallel to the emitted shapes (see
            // `ShapePlan::shared_shapes_after`), so each lands where it was
            // just planned.
            let shape_meshes = plan.shape_meshes(meshes);
            for ((shape, at), mesh) in new_shapes.into_iter().zip(emitted).zip(shape_meshes) {
                prior.by_hash.insert(shape.key, at);
                let Some(rep) = shape.template_of else { continue };
                let bytes = template_bytes(mesh);
                if prior.templates.contains_key(&rep)
                    || self.retained_bytes + bytes > self.template_budget
                {
                    continue;
                }
                self.retained_bytes += bytes;
                prior.templates.insert(rep, RetainedTemplate { mesh: mesh.clone(), at });
            }
        }
        self.vertex_offset = vertex_end;
        self.index_offset = index_end;
        Ok(PlannedBatch {
            mesh,
            vertex,
            index,
            vertex_base,
            index_base,
        })
    }
}

/// The buffers a retained template holds, in bytes.
fn template_bytes(mesh: &MeshData) -> usize {
    (mesh.positions.len() + mesh.normals.len() + mesh.indices.len()) * 4
}

fn overflow() -> ParquetError {
    ParquetError::Overflow("global vertex/index offsets exceed u32".to_string())
}

#[cfg(test)]
#[path = "parquet_stream_shapes_tests.rs"]
mod parquet_stream_shapes_tests;
